File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/ContributorsFragment.java
package openfoodfacts.github.scrachx.openfood.fragments;

import android.os.Bundle;
import android.text.SpannableStringBuilder;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.text.SimpleDateFormat;
import java.util.Date;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.databinding.FragmentContributorsBinding;
import openfoodfacts.github.scrachx.openfood.models.Product;
import openfoodfacts.github.scrachx.openfood.models.State;
import openfoodfacts.github.scrachx.openfood.utils.FragmentUtils;
import openfoodfacts.github.scrachx.openfood.utils.SearchType;
import openfoodfacts.github.scrachx.openfood.views.ProductBrowsingListActivity;

import static android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE;
import static org.apache.commons.lang.StringUtils.isNotBlank;
/*
 * Created by prajwalm on 14/04/18.
 */

/**
 * @see R.layout#fragment_contributors
 */
public class ContributorsFragment extends BaseFragment {
    private FragmentContributorsBinding binding;
    private State stateFromActivity;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentContributorsBinding.inflate(inflater);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        stateFromActivity = FragmentUtils.requireStateFromArguments(this);

        refreshView(stateFromActivity);
    }

    @Override
    public void refreshView(State state) {
        super.refreshView(state);
        stateFromActivity = state;

        final Product product = stateFromActivity.getProduct();
        if (isNotBlank(product.getCreator())) {
            String[] createdDate = getDateTime(product.getCreatedDateTime());
            String creatorTxt = getString(R.string.creator_history, createdDate[0], createdDate[1], product.getCreator());
            binding.creatorTxt.setMovementMethod(LinkMovementMethod.getInstance());
            binding.creatorTxt.setText(creatorTxt);
        } else {
            binding.creatorTxt.setVisibility(View.INVISIBLE);
        }

        if (isNotBlank(product.getLastModifiedBy())) {
            String[] lastEditDate = getDateTime(product.getLastModifiedTime());
            String editorTxt = getString(R.string.last_editor_history, lastEditDate[0], lastEditDate[1], product.getLastModifiedBy());
            binding.lastEditorTxt.setMovementMethod(LinkMovementMethod.getInstance());
            binding.lastEditorTxt.setText(editorTxt);
        } else {
            binding.lastEditorTxt.setVisibility(View.INVISIBLE);
        }

        if (!product.getEditors().isEmpty()) {
            String otherEditorsTxt = getString(R.string.other_editors);
            binding.otherEditorsTxt.setMovementMethod(LinkMovementMethod.getInstance());
            binding.otherEditorsTxt.setText(otherEditorsTxt + " ");
            for (int i = 0; i < product.getEditors().size() - 1; i++) {
                final String editor = product.getEditors().get(i);
                binding.otherEditorsTxt.append(getContributorsTag(editor).subSequence(0, editor.length()));
                binding.otherEditorsTxt.append(", ");
            }
            binding.otherEditorsTxt.append(getContributorsTag(product.getEditors().get(product.getEditors().size() - 1)));
        } else {
            binding.otherEditorsTxt.setVisibility(View.INVISIBLE);
        }

        if (!product.getStatesTags().isEmpty()) {
            binding.statesTxt.setMovementMethod(LinkMovementMethod.getInstance());
            binding.statesTxt.setText("");
            for (int i = 0; i < product.getStatesTags().size(); i++) {
                binding.statesTxt.append(getStatesTag(product.getStatesTags().get(i).split(":")[1]));
                binding.statesTxt.append("\n ");
            }
        }
    }

    /**
     * Get date and time in MMMM dd, yyyy and HH:mm:ss a format
     *
     * @param dateTime date and time in miliseconds
     */
    private String[] getDateTime(String dateTime) {
        long unixSeconds = Long.parseLong(dateTime);
        Date date = new java.util.Date(unixSeconds * 1000L);
        SimpleDateFormat sdf = new java.text.SimpleDateFormat("MMMM dd, yyyy", getResources().getConfiguration().locale);
        SimpleDateFormat sdf2 = new java.text.SimpleDateFormat("HH:mm:ss a", getResources().getConfiguration().locale);
        sdf2.setTimeZone(java.util.TimeZone.getTimeZone("CET"));
        return new String[]{sdf.format(date), sdf2.format(date)};
    }

    private CharSequence getContributorsTag(String contributor) {
        SpannableStringBuilder spannableStringBuilder = new SpannableStringBuilder();
        ClickableSpan clickableSpan = new ClickableSpan() {
            @Override
            public void onClick(@NonNull View view) {
                ProductBrowsingListActivity.startActivity(getContext(), contributor, SearchType.CONTRIBUTOR);
            }
        };
        spannableStringBuilder.append(contributor);
        spannableStringBuilder.setSpan(clickableSpan, 0, spannableStringBuilder.length(), SPAN_EXCLUSIVE_EXCLUSIVE);
        spannableStringBuilder.append(" ");
        return spannableStringBuilder;
    }

    private CharSequence getStatesTag(String state) {
        SpannableStringBuilder spannableStringBuilder = new SpannableStringBuilder();
        ClickableSpan clickableSpan = new ClickableSpan() {
            @Override
            public void onClick(@NonNull View view) {
                ProductBrowsingListActivity.startActivity(getContext(), state, SearchType.STATE);
            }
        };
        spannableStringBuilder.append(state);
        spannableStringBuilder.setSpan(clickableSpan, 0, spannableStringBuilder.length(), SPAN_EXCLUSIVE_EXCLUSIVE);
        spannableStringBuilder.append(" ");
        return spannableStringBuilder;
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/view/contributors/ContributorsFragment.kt
package openfoodfacts.github.scrachx.openfood.features.product.view.contributors

import android.os.Bundle
import android.text.Spanned
import android.text.format.DateFormat
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.text.buildSpannedString
import androidx.fragment.app.viewModels
import dagger.hilt.android.AndroidEntryPoint
import logcat.asLog
import logcat.logcat
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.databinding.FragmentContributorsBinding
import openfoodfacts.github.scrachx.openfood.features.product.edit.ProductEditActivity.Companion.KEY_STATE
import openfoodfacts.github.scrachx.openfood.features.search.ProductSearchActivity.Companion.start
import openfoodfacts.github.scrachx.openfood.features.shared.BaseFragment
import openfoodfacts.github.scrachx.openfood.models.ProductState
import openfoodfacts.github.scrachx.openfood.models.entities.states.StatesName
import openfoodfacts.github.scrachx.openfood.network.ApiFields
import openfoodfacts.github.scrachx.openfood.repositories.TaxonomiesRepository
import openfoodfacts.github.scrachx.openfood.utils.LocaleManager
import openfoodfacts.github.scrachx.openfood.utils.SearchType
import openfoodfacts.github.scrachx.openfood.utils.requireProductState
import java.time.Instant
import java.util.*
import javax.inject.Inject

/**
 * @see R.layout.fragment_contributors
 */

@AndroidEntryPoint
class ContributorsFragment : BaseFragment() {

    private var _binding: FragmentContributorsBinding? = null
    private val binding get() = _binding!!

    private val viewModel: ContributorsViewModel by viewModels()

    @Inject
    lateinit var taxonomiesRepository: TaxonomiesRepository

    @Inject
    lateinit var localeManager: LocaleManager

    private lateinit var productState: ProductState

    private val dateFormat = DateFormat.getDateFormat(requireContext()).apply {
        timeZone = TimeZone.getTimeZone("CET")
    }
    private val timeFormatter = DateFormat.getTimeFormat(requireContext()).apply {
        timeZone = TimeZone.getTimeZone("CET")
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentContributorsBinding.inflate(inflater)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.incompleteStates.setOnClickListener { toggleIncompleteStatesVisibility() }
        binding.completeStates.setOnClickListener { toggleCompleteStatesVisibility() }

        binding.incompleteStates.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_keyboard_arrow_down_grey_24dp, 0)
        binding.completeStates.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_keyboard_arrow_down_grey_24dp, 0)

        refreshView(this.requireProductState())
    }

    override fun refreshView(productState: ProductState) {
        super.refreshView(productState)
        this.productState = productState
        val product = this.productState.product!!

        if (!product.creator.isNullOrBlank()) {
            val createdDate = formatDateTime(product.createdDateTime!!)
            binding.creatorTxt.movementMethod = LinkMovementMethod.getInstance()
            binding.creatorTxt.text = getString(R.string.creator_history, createdDate.first, createdDate.second, product.creator)
        } else {
            binding.creatorTxt.visibility = View.INVISIBLE
        }

        if (!product.lastModifiedBy.isNullOrBlank()) {
            val lastEditDate = formatDateTime(product.lastModifiedTime!!)
            binding.lastEditorTxt.movementMethod = LinkMovementMethod.getInstance()
            binding.lastEditorTxt.text = getString(R.string.last_editor_history, lastEditDate.first, lastEditDate.second, product.lastModifiedBy)
        } else {
            binding.lastEditorTxt.visibility = View.INVISIBLE
        }

        if (product.editors.isNotEmpty()) {
            binding.otherEditorsTxt.movementMethod = LinkMovementMethod.getInstance()
            binding.otherEditorsTxt.text = buildSpannedString {
                append(getString(R.string.other_editors))
                append(" ")
                product.editors.map { getContributorsTag(it).subSequence(0, it.length) }
                    .forEachIndexed { i, el ->
                        if (i > 0) append(", ")
                        append(el)
                    }
                append(getContributorsTag(product.editors.last()))
            }
        } else {
            binding.otherEditorsTxt.visibility = View.INVISIBLE
        }

        // function to show states tags
        viewModel.product.value = product
        viewModel.states.observe(viewLifecycleOwner) { showStatesTags(it) }
    }

    /**
     * Get date and time in MMMM dd, yyyy and HH:mm:ss a format
     *
     * @param epoch unix epoch in seconds
     */
    private fun formatDateTime(epoch: String): Pair<String, String> {
        val date = try {
            Instant.ofEpochSecond(epoch.toLong())
        } catch (err: Exception) {
            logcat { "Could not parse product date $epoch: " + err.asLog() }
            return "" to ""
        }
        return dateFormat.format(date) to timeFormatter.format(date)
    }

    private fun getContributorsTag(contributor: String): CharSequence {
        val clickableSpan = object : ClickableSpan() {
            override fun onClick(view: View) = start(requireContext(), SearchType.CONTRIBUTOR, contributor)
        }
        return buildSpannedString {
            append(contributor)
            setSpan(clickableSpan, 0, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            append(" ")
        }
    }

    private fun getStatesTag(stateName: String, stateTag: String): CharSequence {
        val clickableSpan = object : ClickableSpan() {
            override fun onClick(view: View) = start(requireContext(), SearchType.STATE, stateTag)
        }
        return buildSpannedString {
            append(stateName)
            setSpan(clickableSpan, 0, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        }
    }

    private fun showStatesTags(states: List<StatesName>?) {
        if (states.isNullOrEmpty()) {
            binding.statesTagsCv.visibility = View.GONE
        } else {
            binding.incompleteStatesTxt.movementMethod = LinkMovementMethod.getInstance()
            binding.incompleteStatesTxt.text = ""

            binding.completeStatesTxt.movementMethod = LinkMovementMethod.getInstance()
            binding.completeStatesTxt.text = ""

            states.forEach { state ->
                if (isIncompleteState(state.statesTag)) {
                    binding.incompleteStatesTxt.append(getStatesTag(state.name, state.statesTag.split(":").component2()))
                    binding.incompleteStatesTxt.append("\n")
                } else {
                    binding.completeStatesTxt.append(getStatesTag(state.name, state.statesTag.split(":").component2()))
                    binding.completeStatesTxt.append("\n")
                }
            }
        }

    }


    private fun toggleIncompleteStatesVisibility() {
        if (binding.incompleteStatesTxt.visibility == View.VISIBLE) {
            binding.incompleteStatesTxt.visibility = View.GONE
            binding.incompleteStates.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_keyboard_arrow_down_grey_24dp, 0)
        } else {
            binding.incompleteStatesTxt.visibility = View.VISIBLE
            binding.incompleteStates.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_keyboard_arrow_up_grey_24dp, 0)
        }
    }

    private fun toggleCompleteStatesVisibility() {
        if (binding.completeStatesTxt.visibility == View.VISIBLE) {
            binding.completeStatesTxt.visibility = View.GONE
            binding.completeStates.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_keyboard_arrow_down_grey_24dp, 0)
        } else {
            binding.completeStatesTxt.visibility = View.VISIBLE
            binding.completeStates.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_keyboard_arrow_up_grey_24dp, 0)
        }
    }

    companion object {

        fun newInstance(productState: ProductState) = ContributorsFragment().apply {
            arguments = Bundle().apply {
                putSerializable(KEY_STATE, productState)
            }
        }

        internal fun isIncompleteState(stateTag: String) =
            ApiFields.StateTags.INCOMPLETE_TAGS.any { it in stateTag }
    }
}
